src/Servers/Kestrel/Core/src/Internal/Http2/Http2Stream.cs

```diff
@@ -198,6 +201,15 @@ protected override bool TryParseRequest(ReadResult result, out bool endConnectio
         // do the reading from a pipeline, nor do we use endConnection to report connection-level errors.
         endConnection = !TryValidatePseudoHeaders();
 
+        // 431 if the headers are too large
+        if (TotalParsedHeaderSize > _context.ServiceContext.ServerOptions.Limits.MaxRequestHeadersTotalSize)
```

Comment:
Nit:

```suggestion
        if (TotalParsedHeaderSize > ServerOptions.Limits.MaxRequestHeadersTotalSize)
```

`ServerOptions` is a field, so there's a little less indirection that way.